### ISSUE TITLE
processTSS using large amounts of memory

### ISSUE THREAD

**Issue by rpolicastro** [BrendelGroup/TSRchitect#222]
This report is for for latest github version of TSRchitect.

processTSS seems to be using a large amount of memory. I had 6 fairly small BAM files loaded into a TSRchitect object. When I went to run processTSS using 4 cores I got the following error from the (carbonate) resource manager.

```
=>> PBS: job killed: vmem 101773303808 exceeded limit 34359738368
Error in serialize(data, node$con, xdr = FALSE) :
  Java called System.exit(143) requesting R to quit - trying to recover
```

I had given myself 32GB of memory for my interactive session, but it looks like TSRchitect went up to about 100GB before I got booted.

**Bam file sizes**

```
-rw-r--r-- 1 rpolicas biol  57M Mar 24 19:25 S288C_diamide_1_Aligned.out_cleaned.bam
-rw-r--r-- 1 rpolicas biol  59M Mar 24 19:25 S288C_diamide_2_Aligned.out_cleaned.bam
-rw-r--r-- 1 rpolicas biol  47M Mar 24 19:25 S288C_diamide_3_Aligned.out_cleaned.bam
-rw-r--r-- 1 rpolicas biol  76M Mar 24 19:25 S288C_WT_1_Aligned.out_cleaned.bam
-rw-r--r-- 1 rpolicas biol  86M Mar 24 19:25 S288C_WT_2_Aligned.out_cleaned.bam
-rw-r--r-- 1 rpolicas biol  76M Mar 24 19:25 S288C_WT_3_Aligned.out_cleaned.bam
```

**Session Info**

```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server 7.5 (Maipo)

Matrix products: default
BLAS/LAPACK: /gpfs/home/r/p/rpolicas/Carbonate/.conda/envs/tsrchitect-dev/lib/R/lib/libRblas.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] TSRchitect_1.8.9

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1                    compiler_3.5.1
 [3] BiocManager_1.30.4            later_0.8.0
 [5] GenomeInfoDb_1.18.1           XVector_0.22.0
 [7] AnnotationHub_2.14.2          bitops_1.0-6
 [9] tools_3.5.1                   zlibbioc_1.28.0
[11] digest_0.6.18                 bit_1.1-14
[13] lattice_0.20-38               RSQLite_2.1.1
[15] memoise_1.1.0                 Matrix_1.2-16
[17] DelayedArray_0.8.0            shiny_1.2.0
[19] DBI_1.0.0                     yaml_2.2.0
[21] parallel_3.5.1                rJava_0.9-10
[23] GenomeInfoDbData_1.2.0        rtracklayer_1.42.1
[25] httr_1.4.0                    gtools_3.8.1
[27] XLConnectJars_0.2-15          Biostrings_2.50.2
[29] S4Vectors_0.20.1              IRanges_2.16.0
[31] grid_3.5.1                    stats4_3.5.1
[33] bit64_0.9-7                   Biobase_2.42.0
[35] R6_2.4.0                      AnnotationDbi_1.44.0
[37] XML_3.98-1.19                 BiocParallel_1.16.2
[39] blob_1.1.1                    magrittr_1.5
[41] Rsamtools_1.34.0              matrixStats_0.54.0
[43] promises_1.0.1                htmltools_0.3.6
[45] BiocGenerics_0.28.0           GenomicRanges_1.34.0
[47] GenomicAlignments_1.18.1      XLConnect_0.2-15
[49] SummarizedExperiment_1.12.0   mime_0.6
[51] interactiveDisplayBase_1.20.0 xtable_1.8-3
[53] httpuv_1.5.0                  RCurl_1.95-4.12
```

**Comment by vpbrendel** [comment]
From the information you provided, it's not possible to figure out what went wrong for you.
Could you please run your job within the Singularity Hub TSRchitect container and report whether you see such issues?

**Comment by rpolicastro** [comment]
Looks like running it off the singularity container worked, so the original error was probably caused by something specific to my machine. If nothing else I guess it's a good illustration of the wonders of singularity.

Thanks!

**Comment by rpolicastro** [comment]
Using the R version in the singularity container allowed me to get past processTSS. However, I got a similar error after using determineTSR using 4 cores.

```
... determineTSR ...
=>> PBS: job killed: vmem 104983220224 exceeded limit 68719476736
Error: 'bplapply' receive data failed:
  Java called System.exit(143) requesting R to quit - trying to recover
```

I had given the interactive session 64GB of memory this time, and again it looks like it got to 100GB before it booted me off.

**Session Info**

```
R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] TSRchitect_1.8.9

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1                    compiler_3.5.3
 [3] BiocManager_1.30.4            later_0.8.0
 [5] GenomeInfoDb_1.18.2           XVector_0.22.0
 [7] AnnotationHub_2.14.5          bitops_1.0-6
 [9] tools_3.5.3                   zlibbioc_1.28.0
[11] digest_0.6.18                 bit_1.1-14
[13] lattice_0.20-38               RSQLite_2.1.1
[15] memoise_1.1.0                 Matrix_1.2-16
[17] DelayedArray_0.8.0            shiny_1.2.0
[19] DBI_1.0.0                     yaml_2.2.0
[21] parallel_3.5.3                rJava_0.9-10
[23] GenomeInfoDbData_1.2.0        rtracklayer_1.42.2
[25] httr_1.4.0                    gtools_3.8.1
[27] XLConnectJars_0.2-15          Biostrings_2.50.2
[29] S4Vectors_0.20.1              IRanges_2.16.0
[31] grid_3.5.3                    stats4_3.5.3
[33] bit64_0.9-7                   Biobase_2.42.0
[35] R6_2.4.0                      AnnotationDbi_1.44.0
[37] XML_3.98-1.19                 BiocParallel_1.16.6
[39] blob_1.1.1                    magrittr_1.5
[41] Rsamtools_1.34.1              matrixStats_0.54.0
[43] promises_1.0.1                htmltools_0.3.6
[45] BiocGenerics_0.28.0           GenomicRanges_1.34.0
[47] GenomicAlignments_1.18.1      XLConnect_0.2-15
[49] SummarizedExperiment_1.12.0   mime_0.6
[51] interactiveDisplayBase_1.20.0 xtable_1.8-3
[53] httpuv_1.5.0                  RCurl_1.95-4.12
```

**Comment by vpbrendel** [comment]
Still completely impossible to figure out what was run ... What are the actual R commands you tried to execute?  There is no way TSRchitect should blow up to such huge virtual memory usage given your small input. What happens if you run one input file at a time? What happens if you use 1 processor? Where exactly does the memory hogging happen?

**Comment by vpbrendel** [comment]
The memory hogging was confirmed happening with the old version of loadTSSobj and should have been addressed by pull request #223 .

**Comment by vpbrendel** [comment]
By all my testing, this issue is settled with the latest version on this repository, including the latest version of the TSRchitect container.
